crates/agglayer-interop-grpc-types/src/compat/v1/aggchain_data/v1_to_types.rs

```diff
@@ -0,0 +1,159 @@
+            ),
+            vkey: std::panic::catch_unwind(|| sp1v4_bincode_options().deserialize(&vkey))
+                .map_err(|_| {
+                    Error::deserializing_vkey(Box::new(bincode::ErrorKind::Custom(String::from(
+                        "panic",
+                    ))))
+                })?
+                .map_err(Error::deserializing_vkey)?,
+            version,
+        }))
+    }
+}
+
+impl TryFrom<v1::AggchainData> for AggchainData {
+    type Error = Error;
+
+    fn try_from(value: v1::AggchainData) -> Result<Self, Self::Error> {
+        Ok(match value.data {
+            Some(v1::aggchain_data::Data::Signature(signature)) => AggchainData::ECDSA {
+                signature: (&*signature.value)
+                    .try_into()
+                    .map_err(Error::parsing_signature)?,
+            },
+            Some(v1::aggchain_data::Data::Generic(aggchain_proof)) => {
+                let signature = aggchain_proof
+                    .signature
+                    .as_ref()
+                    .map(|signature| {
+                        (&*signature.value)
+                            .try_into()
+                            .map_err(Error::parsing_signature)
+                    })
+                    .transpose()?
+                    .map(Box::new);
+
+                let AggchainProof {
+                    public_values,
+                    aggchain_params,
+                    proof,
+                } = aggchain_proof.try_into()?;
+
+                AggchainData::Generic {
+                    public_values,
+                    aggchain_params,
+                    signature,
+                    proof,
+                }
+            }
+            Some(v1::aggchain_data::Data::Multisig(multisig)) => {
+                AggchainData::MultisigOnly(multisig.try_into()?)
+            }
+            Some(v1::aggchain_data::Data::MultisigAndAggchainProof(
+                multisig_and_aggchain_proof,
+            )) => AggchainData::MultisigAndAggchainProof {
+                multisig: required_field!(multisig_and_aggchain_proof, multisig),
+                aggchain_proof: required_field!(multisig_and_aggchain_proof, aggchain_proof),
+            },
+            None => return Err(Error::missing_field("data")),
+        })
+    }
+}
+
+impl TryFrom<v1::Multisig> for MultisigPayload {
+    type Error = Error;
+
+    fn try_from(multisig: v1::Multisig) -> Result<Self, Self::Error> {
+        match multisig.data {
+            Some(v1::multisig::Data::Ecdsa(v1::EcdsaMultisig { signatures })) => {
+                if signatures.is_empty() {
+                    return Err(Error::invalid_data(
+                        "Multisig ECDSA doesn't have any signature".to_string(),
+                    ));
+                }
+
+                // Find the maximum key to determine the vector size
+                let max_key = signatures.iter().map(|entry| entry.key).max().unwrap_or(0);
```

Comment:
There is another issue. Since the committee size is not known at this point, we have to infer the vector size from the indices/keys, which is not a trusted input. There should be some max limit. Or switch to a different representation with size proportional to the number of signatures.